packages/ember-application/lib/system/application-instance.js

```diff
@@ -108,26 +118,103 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
     }
   },
 
-  router: computed(function() {
-    return this.lookup('router:main');
-  }).readOnly(),
+  _bootOptions: null,
+  _bootPromise: null,
+  _booted: false,
 
   /**
-    Instantiates and sets up the router, specifically overriding the default
-    location. This is useful for manually starting the app in FastBoot or
-    testing environments, where trying to modify the URL would be
-    inappropriate.
+    Initialize the `Ember.ApplicationInstance` and return a promise that resolves
+    with the instance itself when the boot process is complete.
+
+    The primary task here is to run any registered instance initializers.
 
+    See the documentation on `BootOptions` for the options it takes.
+
+    @private
+    @method boot
     @param options
+    @return {Promise<Ember.ApplicationInstance,Error>}
+  */
+  boot(options) {
+    if (this._bootPromise) { return this._bootPromise; }
+
+    let defer = new Ember.RSVP.defer();
+    let promise = this._bootPromise = defer.promise;
+
+    try {
+      defer.resolve(this._bootSync(options));
+    } catch(e) {
+      defer.reject(e);
+    }
+
+    return promise;
+  },
+
+  /**
+    Unfortunately, a lot of existing code assumes booting an instance is
+    synchronous – specifically, a lot of tests assumes the last call to
+    `app.advanceReadiness()` or `app.reset()` will result in a new instance
+    being fully-booted when the current runloop completes.
+
+    We would like new code (like the `visit` API) to stop making this assumption,
+    so we created the asynchronous version above that returns a promise. But until
+    we have migrated all the code, we would have to expose this method for use
+    *internall* in places where we need to boot an instance synchronously.
+
     @private
   */
-  overrideRouterLocation(options) {
-    var location = options && options.location;
-    var router = get(this, 'router');
+  _bootSync(options) {
+    if (this._booted) { return this; }
```

Comment:
Are there valid use cases for absorbing a double call to `_bootSync`? I can imagine that if the second call has different options (that are thus ignored) and there's no error, tracking that bug down will be extremely annoying.